docs/sdk/src/guides/enable_elastic_scaling_mvp.rs

```diff
@@ -0,0 +1,142 @@
+//! # Enable elastic scaling MVP for a parachain
+//!
+//! **This guide assumes full familiarity with Asynchronous Backing and its terminology, as defined
+//! in <https://wiki.polkadot.network/docs/maintain-guides-async-backing>.
+//! Furthermore, the parachain should have already been upgraded according to the guide.**
+//!
+//! ## Quick introduction to elastic scaling
+//!
+//! [Elastic scaling](https://polkadot.network/blog/elastic-scaling-streamling-growth-on-polkadot)
+//! is a feature that will enable parachains to seamlessly scale up/down their block space usage in
+//! order to increase throughput or lower their latency.
+//!
+//! At present, with Asynchronous Backing enabled, a parachain can only include a block on the relay
+//! chain every 6 seconds, irregardless of how many cores the parachain acquires. Elastic scaling
+//! builds further on the 10x throughput increase of Async Backing, enabling collators to submit up
+//! to 3 parachain blocks per relay chain block, resulting in a further 3x throughput increase.
+//!
+//! ## Current limitations of the MVP
+//!
+//! The full implementation of elastic scaling spans across the entire relay/parachain stack and is
+//! still [work in progress](https://github.com/paritytech/polkadot-sdk/issues/1829).
+//! Below are described the current limitations of the MVP:
+//!
+//! 1. **Limited core count**. Parachain block authoring is sequential, so the second block will
```

Comment:
How do we know that these 3 para-blocks are still valid when imported in 3 parallel cores?

For example, there are 2 tx in each parablock. The collator proposes `[t1, t2, t3, t4, t5, t6]` and they are all valid. But the validity of `t6` depends on the execution of `t1`. When imported in 3 cores, `t1` and `t6` are no longer present. 

In general, I would assume all of this to be fixed in the cumulus block building code. My question is, does it?

Reply:
These 3 blocks are expected to form a chain, the ones that don't will not be included.

Reply:
> These 3 blocks are expected to form a chain, the ones that don't will not be included.

yes, also a candidate will not be included until all of its ancestors are included. if one ancestor is not included (times out availability) or is concluded invalid via a dispute, all of its descendants will also be evicted from the cores. So we only deal with candidate chains

Reply:
Sorry, I still don't get this. 

@sandreim if they form a chain, and part of the chain is executed in one core and part of it in another core, how does either of the cores check that the whole thing is a chain? 

in my example, `[t1, t2, t3, t4, t5, t6]`, `[t1, t2, t3]` goes into one core, `[t4, t5, t6]` into another. The whole `[t1 -> t6]` indeed forms a chain, and execution of t5 depends on the execution of t2. 

Perhaps what you mean to say is that the transactions that go into different cores must in fact be independent of one another?

Reply:
The transactions are not independent. We achieve parallel execution even in that case, and still check they form a chain by passing in the appropriate validation inputs (https://github.com/paritytech/polkadot-sdk/blob/d1979d4be9c44d06c62aeea44d30f5f201d67b5e/polkadot/primitives/src/v7/mod.rs#L661) . We can validate t2 because we already have the parent head data of t1 from the collator of t2. So we can correctly construct the inputs and the PoV contains the right data ( t2 was built after t1 by the collator).

Does it makes sense.l?

Reply:
> [t1, t2, t3] goes into one core, [t4, t5, t6] into another.

This is not really possible. If a parachain uses multiple cores, all the blocks occupying the cores will form a chain.
In your example, on core 1, we'll probably have [t1,t3,t5] and on core 2 [t2,t4, t6].

Reply:
> > [t1, t2, t3] goes into one core, [t4, t5, t6] into another.
> 
> This is not really possible. If a parachain uses multiple cores, all the blocks occupying the cores will form a chain. In your example, on core 1, we'll probably have [t1,t3,t5] and on core 2 [t2,t4, t6].

I don't think you can generalize, because you are assuming everyone will use a specific collator implementation, right? I have leases two cores, and I run their collators, and therefore it is me who chose to which cores I send which transactions. 

Let's speak only from the perspective of Polkadot, and its cores here. For all I know, collators are written by arbitrary external teams, in much the same way that ETH block builders are separate from core validators. 

From the perspective of Polkadot, when the same parachain is using two cores, what is the assumptions?

> We can validate t2 because we already have the parent head data of t1 from the collator of t2. So we can correctly construct the inputs and the PoV contains the right data ( t2 was built after t1 by the collator).

You are also talking about the collators here, while I am curious to know how does two different Polkadot cores (and therefore two different backing groups) simultaneously validate two blocks that might have inter-dependence.

Reply:
Is this the answer? 

`[t1, t2, t3]` goes into one core, `[t4, t5, t6]`, but the PoV of the latter contains the full execution of the former? 

I think this is fine, but truthfully to scale up, I think different transactions going into different cores must be independent, or else the system can only scale as much as you can jack up one collator.

Reply:
> but the PoV of the latter contains the full execution of the former?

PoV of `[t4, t5, t6]` would refer to the state post `[t1, t2, t3]` execution.

> I think different transactions going into different cores must be independent, or else the system can only scale as much as you can jack up one collator

One way of achieving that would be to have a DAG instead of a blockchain (two blocks having the same parent state). But then you'd need to somehow ensure they are truly independent. This could be done with e.g. specifying dependencies in the transactions themselves (a ala Solana or Ethereum access lists).

Another way would be to implement a collator differently with optimistic concurrency control (ala Monad). This only requires modification on the collator side and does not affect transaction format.

Reply:
Okay, thanks @ordian. 

I totally agree with all of your directions as well. I am not sure if you have seen it or not, but my MSC Thesis was on the same topic 🙈 https://github.com/kianenigma/SonicChain. I think what I have done here is similar to access list, and it should be quite easy to add to FRAME and Substrate: each tx to declare, via its code author, what storage keys it "thinks" it will access. Then the collators can easily agree among themselves to collate non-conflicting transactions. 

This is a problem that is best solve from the collator side, and once there is a lot of demand. Polkadot is already doing what it should do, and should not do any "magic" to handle this. 

Once there is more demand: 

1. Either collators just jack up, as they kinda are expected to do now. This won't scale a lot but it will for a bit. 
2. I think the access list stuff is super cool and will scale
3. OCC is fancy but similarly doesn't scale, because there is only so many CPU cores, and you are still bound to one collator somehow filling up 8 Polkadot cores. Option 2 is much more powerful, because you can enable 8 collators to fill 8 blocks simultaneously.

Reply:
Current status quo is that we rely on 1 (beefy collators). 2 for sure is something that can scale well, but it seems to be complicated and is not really compatible with the relay chain which expects chains not a DAG. https://github.com/paritytech/polkadot-sdk/issues/4696#issuecomment-2173250798 shows how the limitations of what is possible with ref hw and 5 collators.

We did a nice brainstorming session with @skunert and @eskimor on the subject some time ago. We think that best way to go forward is to implement a transaction streaming mechanism. At the begging of each slot, the block author sends transactions to the next block author as it pushes them in the current block. By the time it announces the block, the next author should already have all state changes applied and doesn't need to wait to import it and can immediately start building his own block. And so on. 

If that is not enough, next block author can start to speculatively build the it's next block and add/remove transactions as it learns what the current author is putting in his blocks.